### ISSUE TITLE
Recreate Domain Theme

### ISSUE THREAD

**Issue by agentrickard** [agentrickard/domain#102]
With the new config system, this might be part of Domain Config. Investigate and make a plan.

**Comment by agentrickard** [comment]
Theme settings are handled by the config system. E.g.

* system.theme
* system.theme.global
* THEME_NAME.settings
* color.theme.THEME_NAME

Which means we handle them via Domain Config, and build a UI in Domain Theme, I think.

**Comment by ashishdeynap** [comment]
Hey Rickard,

Appreciate your efforts in migrating this Module to version 8. However, Is there any update in Domain Theme module for D8 or can you suggest me a way to tackle this problem.

Any response would be welcomed

**Comment by agentrickard** [comment]
I haven't done any research beyond what is listed here. I would start with the following:

* Try using Domain Config to load domain-specific theme settings.
* Document here what works and what fails.

After that, we can investigate if building a UI is practical. Right now, it is _not_ practical for Domain Config.

**Comment by agentrickard** [comment]
Theme settings are just configuration yml files now. So you can export a YML file, rename it 'domain.config.[DOMAIN_MACHINE_NAME].[CONFIG.NAME].yml' edit it, and then import to your site.

There is no UI for that, and we'd like to know what breaks when you try that. I would start with system.theme and create a file like so:

domain.config.example_com.system.theme.yml
```
admin: seven // set this to something other than your main site.
default: bartik // this too!
_core:
  default_config_hash: fOjer9hADYYnbCJVZMFZIIM1azTFWyg84ZkFDHfAbUg
```

**Comment by zerolab** [comment]
We use that on a live site with 2 domains.
Each domain has a different theme and the system.theme.yml for each of them specifies a different theme.

It works well, but did lead to several moments of confusion for other developers when trying to change some other properties. A domain-sensitive UI would have made it easier for sure

**Comment by agentrickard** [comment]
Right, I just have no idea how we might do that, given how other configuration forms and schema are handled.

Good to know it "works" -- have you tried other theme settings files?

**Comment by zerolab** [comment]
Here is the list of domain settings files currently in use:

```
domain.config.mysite.smtp.settings.yml
domain.config.mysite.system.site.yml
domain.config.mysite.system.theme.yml
domain.config.mysite_subdomain.smtp.settings.yml
domain.config.mysite_subdomain.system.site.yml
domain.config.mysite_subdomain.system.theme.yml
domain.config.mysite_subdomain.user.settings.yml
````

**Comment by agentrickard** [comment]
I'd love to see a little bit of documentation about these and if they cause any issues.

We could stub out Domain Theme module with just a README for now with that.

**Comment by nedjo** [comment]
Semi related note.

I tried overriding color module settings, trying to have the same theme but different colour settings per domain, but ran into a caching issue in that the first page hit, on any domain, set the colours for all pages on all domains.

**Comment by agentrickard** [comment]
We had similar problems in Drupal 7. Theme settings are a pain. To me, that's a core caching bug.

I wonder if cache tags and default cache contexts would help with that case.

**Comment by agentrickard** [comment]
e.g. in services.yml:

```
required_cache_contexts: ['languages:language_interface', 'theme', 'user.permissions']
```

Might need to be:

```
required_cache_contexts: ['languages:language_interface', 'theme', 'user.permissions', 'url.site']
```

**Comment by Steelhammer100** [comment]
Where do I have to put the domain.config.example_com.system.theme.yml file? In Configuration Manager there is only Domain Record and Domain Alias for Importing.

**Comment by agentrickard** [comment]
That would be considered "Simple configuration." 

We should document that you cannot set Domain Records or Aliases per domain (that causes an infinite loop).

All other types of config map to the schema definition of the settings that you are overriding.

**Comment by 4ndir** [comment]
I just tried a different theme per domain, which works fine and which is great! However, the same theme with different settings does not work. 

This does not help either:
`required_cache_contexts: ['languages:language_interface', 'theme', 'user.permissions', 'url.site']`

**Comment by agentrickard** [comment]
"Does not work" is not a useful report. Please be more specific about expected and observed behaviors. We'd also need to see what configuration you tried -- the actual YML file.

**Comment by 4ndir** [comment]
Sure! I also did some more testing. I have two domains: 
1. domain1-da.test
2. domain2-da.test

Setting the theme to bartik and loading another logo for domain2 works indeed fine.
`domain.config.domain2_da_test.bartik.settings`
```
logo:
  path: sites/default/files/bilder/custom.jpg
```

Setting the theme to a custom theme (subtheme of bootstrap) and editing properties of domain2 leads to some weird behaviour:
domain2
`domain.config.domain2_da_test.customtheme.settings`
`cdn_jsdelivr_theme: cyborg
navbar_position: fixed-top`

domain1
`customtheme.settings` 
`cdn_jsdelivr_theme: readable
navbar_position: fixed-top`

Now, if I clear caches on domain1, the settings for domain1 apply for domain1 and domain2. If I visit domain2 and clear caches, settings for domain2 apply for domain1 and domain2.

**Comment by 4ndir** [comment]
The same happens if I override theme settings for domain1 as well:
`domain.config.domain1_da_test.customtheme.settings`
```
cdn_jsdelivr_theme: readable
navbar_position: fixed-top
```

**Comment by agentrickard** [comment]
That is likely a deeply nested core caching issue. If you can find out where the theme settings are getting cached, we may be able to work around that.

**Comment by webhype** [comment]
I was able to change the logo using logo: and path:, bit striking out on all other theme properties. Can you post a 'for dummies' example for changing the site name and slogan? Sorry if this is a stupid question, I am not a theme autor and never intended to go that deep, but I am willing to upload some custom configurations, if I can find documentation. If this is the wrong place to ask for help, please advise!

**Comment by webhype** [comment]
Figured it out, I was trying to override the wrong item. I uploaded a single configuration item at /admin/config/development/configuration/single/import > "Simple Configuration" > "domain.config.example_com.system.site" with the following content: 

name: 'Example Inc.'
slogan: 'Home of the best examples'

After clearing cache, it displayed that at my domain with machine name 'example_com'. 

Thanks for your hard work on this module! :)

**Comment by agentrickard** [comment]
That's great. This is exactly the right place to ask questions. I honestly haven't had time to test any of the theme handling stuff.

If you wrote up a bit of documentation here, I can add it to the module.

**Comment by webhype** [comment]
Well the documentation at https://github.com/agentrickard/domain/tree/8.x-1.x/domain_config already said it all, I just discovered the page after I had figured it out myself lol.

Like nedjo earlier, I am hitting a snag now with the Color module, trying to give a different color palette to a different domain. I followed the "Domain" logic and created a .yml configuration named "domain.config.example_com.color.theme.theme_machine_name" with contents like this:

palette:
  title: '#8d6f54'
  text: '#888888'
  metatext: '#777777'
.... and so on...

However, it seems to do absolutely nothing. At least nedjo saw a color change (though the same on all domains). I got no effect. I uploaded system.yml to /sites/example_com/ (is that the right place?) with the added caching directive ('url.site'); seemed to help nothing. 

Any new insights on Domain & Color module since July? Thanks Rickard!

**Comment by agentrickard** [comment]
No insights. Domain Theme is pretty low on my priority list.

It's not system.yml, it's `services.yml` see /sites/default/default.services.yml`, so the new file would be `/sites/default/services.yml.`

**Comment by webhype** [comment]
Yes, services.yml, typo. And yes I had that file in that location also, didn't seem to make a difference. Thanks for trying to help. If someone figures out how to change the color palette by domain, please post the solution here! Thanks!

**Comment by zerolab** [comment]
have you tried doing it in your local settings.php?
you will need to override the whole `color.theme.theme_machine_name.yml`.

Something like 

```
$config['domain.config.example_com.color.theme.theme_machine_name'] = [
  'palette' => [
     'title' => '#8d6f54', 
  ],
  // etc.
];
```

Note that this requires a recent dev version

**Comment by webhype** [comment]
Hi thanks Zerolab, unfortunately it also has no effect. If by "recent dev version" you mean the version of the Domain module, I used the one from 07-DEC-2016. I did not use the the latest Drupal dev version, if you meant that. I used the latest release version.

In the meantime I had my developer investigate and he got it to work, but not without hacking the Color module and two files in core/lib/Drupal/Core/Asset :

LibraryDiscoveryCollector.php
AssetResolver.php

The problem seems to be that D8 has no notion of needing to cache different CSS files for different domains (or FQDNs to be exact). So he had to "teach" D8 to distinguish caches by domain, as I understand it. 

I can elaborate more if there is interest. We can now switch between color palettes with a (hacked) domain dropdown right in the Color module screen.

Where to go from here? Perhaps we should suggest to the D8 developers to make the cache FQDN-sensitive? Unfortunately we don't have time to contribute polished code to them directly and become part of the D8 project, but if we got the attention of whoever works on caching, perhaps we can forward our hack and see if the principles behind it could be implemented in the next core release. Any pointers where to go?

**Comment by agentrickard** [comment]
I'd really like to see the code the developer ended up using. As far as I understand it, caching by FQDN is a core feature -- it's just that we found some areas where it isn't used.

In the case of home-page per domain, we had to override the core route caching service (see `domain_config/src/Routing/DomainRouteProvider.php`.

We may be able to do something similar, since the LibraryDiscovery class is a service:

```
  library.discovery:
    class: Drupal\Core\Asset\LibraryDiscovery
    arguments: ['@library.discovery.collector']
```

As is the asset.resolver.

```
  asset.resolver:
    class: Drupal\Core\Asset\AssetResolver
    arguments: ['@library.discovery', '@library.dependency_resolver', '@module_handler', '@theme.manager', '@language_manager', '@cache.data']
```

Have your developer look at how we modify the core service inside of ```domain_config/src/DomainConfigServiceProvider.php```

We could use the same approach to extend / override these two classes cleanly, without patching core.

Then we can file a core issue that shows why we had to do that and suggest a patch.

**Comment by agentrickard** [comment]
Here's the problem pretty clearly -- there is no expectation that the same theme might have different color / css files.

From `AssetResolver::getCssAssets()`

```
    $cid = 'css:' . $theme_info->getName() . ':' . Crypt::hashBase64(serialize($libraries_to_load)) . (int) $optimize;
    if ($cached = $this->cache->get($cid)) {
      return $cached->data;
    }
```

The same is true in the JS handler.

It is not immediately clear what alterations need to be made to the LibraryDiscovery class.

**Comment by webhype** [comment]
Our hack addresses that line exactly (line 118 in AssetResolver.php) in getCssAssets():

`    if ( (\Drupal::moduleHandler()->moduleExists('domain') && ($active_domain = \Drupal::service('domain.negotiator')->getActiveDomain()))) {
        $domain_id = $active_domain->get('id');
        $cid = 'css:' . $theme_info->getName() . '_' . $domain_id . ':' . Crypt::hashBase64(serialize($libraries_to_load)) . (int) $optimize;
    }else{
        $cid = 'css:' . $theme_info->getName() . ':' . Crypt::hashBase64(serialize($libraries_to_load)) . (int) $optimize;
    }`

In LibraryDiscoveryCollector.php we hacked getCid() around line 59 or so:

`protected function getCid() {
// ... unpatched stuff ...
    if ((\Drupal::moduleHandler()->moduleExists('domain') && ($active_domain = \Drupal::service('domain.negotiator')->getActiveDomain()))) 
            $this->cid .= ':' . $active_domain->get('id')); 

    return $this->cid;
  }`

There are too many patches in color.module, but it basically revolves around inventing our own color key that includes the "active" domain as distinguishing factor -- if interested, where can I upload it so it gets to you Rickard?

**Comment by agentrickard** [comment]
You could upload to a Gist or email it to me (agentrickard) at Gmail.

**Comment by agentrickard** [comment]
Oh, and show your dev how to override services to avoid hacking core :-)

**Comment by webhype** [comment]
Emailed :) Thanks!

**Comment by andypost** [comment]
there's sandbox code https://www.drupal.org/sandbox/aloknarwaria/2691703

**Comment by nedjo** [comment]
If you're using Domain Config overrides for the default theme, you can end up with unexpected results in admin paths that include the theme.

The path `admin/structure/block` is for managing blocks of the default theme--which changes per domain.

Say you have a default domain, Domain A, and the default theme value is "Theme A". On the non-default domain Domain B you've saved an overridden default theme value of "Theme B".

The path `admin/structure/block` will always have the same label, but will lead to a different set of blocks if visited from different domains.

So if you're clearing caches with for example the [Admin Toolbar](https://www.drupal.org/project/admin_toolbar), you need to take care to do so from the default domain, where the default theme is not set to an overridden value.

If you clear caches from Domain B, the path `admin/structure/block` will end up with the label "Theme B". If you visit the path on Domain A, it will appear to be displaying the blocks of "Theme B" but will actually display the blocks of "Theme A", and "Theme B" blocks will be inaccessible. Conversely, if you clear caches on Domain A and then visit the path on Domain B, it will have the label "Theme A" but will actually display blocks of "Theme B".

 The best practice is probably to:
* Always clear caches from the default domain.
* Always manage blocks from the default domain.

**Comment by nedjo** [comment]
Filed a related core bug, [Router path admin/structure/block should use default theme without overrides](https://www.drupal.org/node/2844784).

**Comment by cmodern** [comment]
I'm trying to upload different logo images for each domain on D8, am wondering if you might guide me on how to do that with code, given the current status of the D8 Domain Theme module? It was easy to do in D7 domain access by going to each domain's theme settings, but I understand the UX for this is not ready yet for the D8 version of domain access. What would be needed to create that functionality using code? Thanks in advance.

**Comment by nedjo** [comment]
@cmodern 

See the [README file for the Domain Config module](https://github.com/agentrickard/domain/blob/8.x-1.x/domain_config/README.md).

**Comment by axel-rutz** [comment]
So essentially, domain_theme module is not needed as this can be done with domain_config.

So proposing to close this for more specific issues. I'd propose:

* Leverage configelement.module for a domain_config_ui #452
* Add domain_assets.module #453
* [Color css files are not regenerated when deploying through configuration management \[\#2995825\] \| Drupal\.org](https://www.drupal.org/project/drupal/issues/2995825#comment-12832872) (needs some love to fix tests and set RTBC)

The second and third of which allows us to have color.module per domain✌ on a drutopia-based site which will go into production this weekend.

**Comment by agentrickard** [comment]
Take a look at what's going on over at #434 which gives us a good run at making config manageable through the UI. I was thinking about leveraging some of that for Domain Theme.

**Comment by leo72** [comment]
Going to try this module. Very promissing.
